File: src/main/java/fields/pin/PinFields.java
package fields.pin;

import fields.BaseFields;
import fields.CreatorFields;

import java.util.ArrayList;
import java.util.List;

import static fields.FieldSerializer.serialize;
import static org.apache.commons.lang3.StringUtils.isNotBlank;
import static org.apache.commons.lang3.StringUtils.join;

public class PinFields extends BaseFields {
    private final CreatorFields creatorFields = new CreatorFields();
    private final BoardFields boardFields = new BoardFields();

    @Override
    public String build() {
        final String serializedCreatorFields = creatorFields.build();
        final String serializedBoardFields = boardFields.build();
        final String serializedPinFields = serialize(fields);
        final List<String> serializedFieldsList = new ArrayList<>();

        if (isNotBlank(serializedCreatorFields)) {
            serializedFieldsList.add("creator(" + serializedCreatorFields + ")");
        }
        if (isNotBlank(serializedBoardFields)) {
            serializedFieldsList.add("board(" + serializedBoardFields + ")");
        }
        if (isNotBlank(serializedPinFields)) {
            serializedFieldsList.add(serializedPinFields);
        }

        return join(serializedFieldsList, ",");
    }

    // TODO:  is there a better way to do this? i.e., formulaically
    @Override
    public PinFields setAll() {
        this.setCounts().setLink().setMetadata().setNote().setUrl();
        creatorFields.setAll();
        boardFields.setAll();
        return this;
    }

    public CreatorFields getCreatorFields() {
        return creatorFields;
    }

    public BoardFields getBoardFields() {
        return boardFields;
    }

    public PinFields setLink() {
        this.fields.add("link");
		return this;
    }

    public PinFields setCounts() {
        this.fields.add("counts");
		return this;
    }

    public PinFields setNote() {
        this.fields.add("note");
		return this;
    }

    public PinFields setUrl() {
        this.fields.add("url");
        return this;
    }

    public PinFields setMetadata() {
        this.fields.add("metadata");
        return this;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        PinFields pinFields = (PinFields) o;

        if (creatorFields != null ? !creatorFields.equals(pinFields.creatorFields) : pinFields.creatorFields != null)
            return false;
        if (boardFields != null ? !boardFields.equals(pinFields.boardFields) : pinFields.boardFields != null)
            return false;
        return !(fields != null ? !fields.equals(pinFields.fields) : pinFields.fields != null);

    }

    @Override
    public int hashCode() {
        int result = creatorFields != null ? creatorFields.hashCode() : 0;
        result = 31 * result + (boardFields != null ? boardFields.hashCode() : 0);
        result = 31 * result + (fields != null ? fields.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return "PinFields{" +
                "creatorFields=" + creatorFields +
                ", boardFields=" + boardFields +
                ", fields=" + fields +
                '}';
    }
}

File: src/main/java/com/chrisdempewolf/pinterest/fields/pin/PinFields.kt
package com.chrisdempewolf.pinterest.fields.pin

import com.chrisdempewolf.pinterest.fields.BaseFields
import com.chrisdempewolf.pinterest.fields.CreatorFields

import java.util.ArrayList

import com.chrisdempewolf.pinterest.fields.FieldSerializer.serialize
import com.chrisdempewolf.pinterest.fields.Fields
import org.apache.commons.lang3.StringUtils.isNotBlank
import org.apache.commons.lang3.StringUtils.join

class PinFields : BaseFields() {
    val creatorFields = CreatorFields()
    val boardPinFields = BoardPinFields()

    override fun build(): String {
        val serializedCreatorFields = creatorFields.build()
        val serializedBoardFields = boardPinFields.build()
        val serializedPinFields = serialize(fields)
        val serializedFieldsList = ArrayList<String>()

        if (isNotBlank(serializedCreatorFields)) {
            serializedFieldsList.add("creator($serializedCreatorFields)")
        }
        if (isNotBlank(serializedBoardFields)) {
            serializedFieldsList.add("board($serializedBoardFields)")
        }
        if (isNotBlank(serializedPinFields)) {
            serializedFieldsList.add(serializedPinFields)
        }

        return join(serializedFieldsList, ",")
    }

    override fun withAll(): PinFields {
        withCounts()
                .withLink()
                .withMetadata()
                .withNote()
                .withURL()
                .withColor()
                .withAttribution()
                .withCreatedAt()
        creatorFields.withAll()
        boardPinFields.withAll()
        return this
    }

    fun withLink(): PinFields {
        fields = fields.plus("link")
        return this
    }

    fun withCounts(): PinFields {
        fields = fields.plus("counts")
        return this
    }

    fun withNote(): PinFields {
        fields = fields.plus("note")
        return this
    }

    fun withURL(): PinFields {
        fields = fields.plus("url")
        return this
    }

    fun withMetadata(): PinFields {
        fields = fields.plus("metadata")
        return this
    }

    fun withColor(): PinFields {
        fields = fields.plus("color")
        return this
    }

    fun withAttribution(): PinFields {
        fields = fields.plus("attribution")
        return this
    }

    fun withCreatedAt(): PinFields {
        fields = fields.plus("created_at")
        return this
    }

    fun with(field: String): PinFields {
        fields = fields.plus(field)
        return this
    }
}
